blog/_posts/2021-07-08-portforwarding-is-awesome.md

```diff
@@ -0,0 +1,292 @@
+Accept: */*\r
+\r
+< 2021/07/07 13:18:16.172097  length=356 from=0 to=355
+HTTP/1.1 200 OK \r
+```
+
+A `socat` process can intercept lots of different types of connections! For
+example, the Docker Daemon listens on a unix socket
+`/var/run/docker.sock`. `socat` is a great tool for seeing how the Docker CLI
+interacts with the Docker Service!
+
+```
+sudo mv /var/run/docker.sock /var/run/docker.sock.original
+sudo socat -v UNIX-LISTEN:/var/run/docker.sock,mode=777,reuseaddr,fork UNIX-CONNECT:/var/run/docker.sock.original
+```
+
+Now, when I run a Docker CLI command, I can see which HTTP requests it makes:
+
+```
+# docker ps
+...
+GET /v1.41/containers/json HTTP/1.1\r
+Host: docker\r
+User-Agent: Docker-Client/20.10.7 (linux)\r
+\r
+< 2021/07/07 13:40:15.162007  length=6647 from=280 to=6926
+HTTP/1.1 200 OK\r
+...
+```
+
+Remember to restore the socket when you're done!
+
+```
+sudo mv /var/run/docker.sock.original /var/run/docker.sock
+```
+
+## Connecting to multi-service apps with `kubectl port-forward`
+
+Pretty much every time I do development in Kubernetes, I use `kubectl
+port-forward`. You can think of it as a chain of `socat` commands that copy socket data
+from a local port (`localhost:8080`) to the Kubernetes API server, then to the inside of
+a Pod running in Kubernetes.
+
+In this post, I'm mostly going to focus on how we use `port-forward` in dev. If
+you want to learn more about how it works under the covers, Ivan Sim wrote [a
+deep dive](https://itnext.io/how-it-works-kubectl-exec-e31325daa910) into
+`kubectl exec` that I like. Even though it's about `exec`, not `port-forward`,
+they follow a similar flow.
+
+The simplest way to use `kubectl port-forward` is to connect to a pod:
+
+```
+kubectl port-forward example-go-7ff5965bbb-6jblq 8080:8000
+```
+
+This copies socket data from `localhost:8080` to port 8000 inside the pod `example-go-7ff5965bbb-6jblq`.
+
+But we usually don't deal directly with Kubernetes pods! We deal with
+Deployments / StatefulSets (resources that manage pods) or Services / Ingresses
+(resources that redirect traffic to pods).
+
+Fortunately, `kubectl port-forward` has some nice utilities that infer pods from deployments and services:
+
+```
+kubectl port-forward deployment/example-go 8080:8000
+```
+
+But beware! This doesn't behave like most people expect! This uses the
+Deployment to choose a single pod.  If that pod is deleted, the port-forward
+breaks and emits errors until you restart it (which will force it to choose a
+new pod).
+
+Meanwhile, the borked port-forward is using up a terminal.
+
+Can we do better?
+
+## Inside the Box: A port-forward as Just Another Server
+
+One way to think about `kubectl port-forward` is just another server. It's a very simple server!
+But it tastes great with lots of other tooling that we use to manage multiple servers and auto-restart
+them when they're not healthy.
+
+For example, I might use `entr` to restart the port-forward when I hit spacebar.
```

Comment:
I don't understand how spacebar does anything here? 🤔

Reply:
From the `entr` manual:

```
     entr listens for keyboard input and responds to the following commands:

     ⟨space⟩  Execute the utility immediately.  If the -r option is set this
              will terminate and restart the child process as if a file change
              event had occurred.
```

Reply:
Oh, whoa. TIL.